controllers/datadogagent/controller_reconcile_agent.go

```diff
@@ -238,3 +240,79 @@ func (r *Reconciler) generateNodeAffinity(p string, affinity *corev1.Affinity) *
 
 	return affinity
 }
+
+func (r *Reconciler) handleProviders(ctx context.Context) (*map[string]struct{}, error) {
+	if err := r.updateProviderStore(ctx); err != nil {
+		return nil, err
+	}
+
+	if err := r.cleanupUnusedProviders(ctx); err != nil {
+		return nil, err
+	}
+
+	return r.providerStore.GetProviders(), nil
+}
+
+// updateProviderStore recomputes the required providers by making a
+// call to the apiserver for an updated list of nodes
+func (r *Reconciler) updateProviderStore(ctx context.Context) error {
+	nodeList := corev1.NodeList{}
+	err := r.client.List(ctx, &nodeList)
+	if err != nil {
+		return err
+	}
+
+	// recompute providers using updated node list
+	providersList := make(map[string]struct{})
+	for _, node := range nodeList.Items {
+		provider := kubernetes.DetermineProvider(node.Labels)
+		if _, ok := providersList[provider]; !ok {
+			providersList[provider] = struct{}{}
+			r.log.V(1).Info("New provider detected", "provider", provider)
+		}
+	}
+
+	r.providerStore.Reset(providersList)
+
+	return nil
+}
+
+// cleanupUnusedProviders deletes ds/eds from providers that are not
+// present in the provider store. If there are no providers in the
+// provider store, do not delete any ds/eds since that would delete
+// all node agents.
+func (r *Reconciler) cleanupUnusedProviders(ctx context.Context) error {
+	if r.options.ExtendedDaemonsetOptions.Enabled {
+		edsList := edsv1alpha1.ExtendedDaemonSetList{}
+		if err := r.client.List(ctx, &edsList, client.HasLabels{apicommon.MD5AgentDeploymentProviderLabelKey}); err != nil {
+			return err
+		}
+
+		for _, eds := range edsList.Items {
+			provider := eds.Labels[apicommon.MD5AgentDeploymentProviderLabelKey]
+			if len(*r.providerStore.GetProviders()) > 0 && !r.providerStore.IsPresent(provider) {
+				if err := r.client.Delete(ctx, &eds); err != nil {
```

Comment:
Have you considered using `cleanupV2(Extended)DaemonSet`? it handles error a bit differently by `return reconcile.Result{}, err` instead of  `return nil, err` and also records events.

Reply:
Looking a bit more at this it seems we have two branches which deal with DS/EDS cleanup. One `reconcileInstanceV2 -> handleProviders -> cleanupUnusedProviders` and another `reconcileInstanceV2 -> reconcileV2Agent -> cleanupV2ExtendedDaemonSet`.

I wonder if it would be better to utilize same code path for provider DS cleanup as for a regular DS/EDS. That would be a bit tricky to implement though.

Reply:
i agree that it would be nice to have one cleanup method. if it will be tricky to implement though maybe we can make a separate PR to tackle that

Reply:
Created a card for this for our backlog